### ISSUE TITLE
batch_norm doesn't bump version counter in running stats

### ISSUE THREAD

**Issue by SsnL** [pytorch/pytorch#13402]
This can cause correctness problems if you mix `eval` and `training` modes because `eval` mode backward needs the running stats, and `training` mode forward updates the running stats, even if the `training` mode result is unused. E.g.,

```py
import torch

bn = torch.nn.BatchNorm2d(3).double()
x = torch.randn(2, 3, 1, 1).double().requires_grad_()


def f(x):
  bn.running_mean.data.fill_(0)
  bn.running_var.data.fill_(1)
  bn.eval(); y0 = bn(x)
  bn.train(); unused = bn(x)  # running buffers are changed in-place here
  return y0

torch.autograd.gradcheck(f, (x,))  # errors

def g(x):
  bn.running_mean.data.fill_(0)
  bn.running_var.data.fill_(1)
  bn.eval(); y0 = bn(x)
  bn.running_mean = bn.running_mean.clone()
  bn.running_var = bn.running_var.clone()
  bn.train(); unused = bn(x)
  return y0

torch.autograd.gradcheck(g, (x,))  # passes

```

cc @colesbury

**Comment by v0dro** [comment]
Hi Sameer here from Quansight. Can you assign me to this issue?

**Comment by v0dro** [comment]
So if my understanding of the issue is correct, the problem is that `bn(x)` is supposed to update the version counter of `bn.running_mean` and `bn.running_var` after performing in-place operations on them since performing in-place operations involves bumping the version counter so that the computation graph can track that a tensor that it is using is old.

The fix will basically involve changing the place within `Module.__call__()` which is not currently updating the version counter in train mode to do so.

Is that correct? I'll also post exactly how I plan to solve this problem soon.

**Comment by albanD** [comment]
Actually I think the problem is with the automatically generated code:
The main batchnorm declaration can be found [here](https://github.com/pytorch/pytorch/blob/0d22f3b1709467ee42c5b00628ad5648c2b27fb6/tools/autograd/derivatives.yaml#L582). As you can see, the stats are marked as optional with the `?`.
The problem is that they are not marked as being changed inplace. This is done with an `!` as is done [here](https://github.com/pytorch/pytorch/blob/0d22f3b1709467ee42c5b00628ad5648c2b27fb6/tools/autograd/derivatives.yaml#L395) for example for index_add_.

This definitely looks like the root cause of this issue, I am not sure why it is this way though. In particular, I am not sure optional arguments can be marked as being modified inplace (I couldn't find one by doing a quick search) with the current code generation we have.

**Comment by gchanan** [comment]
I don't think this issue is small -- fixing it properly probably involves fixing a fair amount of codegen.

**Comment by jjsjann123** [comment]
Inplace operation on module parameters would be bad for JIT. Due to the aliasing assumption, any fusion / re-ordering across the node would not be possible.
The proper thing to do would be to 1) update annotation with in-place marker and 2) properly pass aliasing information to compiled method, but we don't have tools to do 2) yet.

I would suggest that we remove the in-place operation here and have primitive kernel to update running_mean/running_var.

**Comment by v0dro** [comment]
So there has been quite a bit of discussion on this issue on the Pytorch [slack channel](https://pytorch.slack.com/archives/C3PDTEV8E/p1571092762035100), and from the looks of it this issue definitely does not seem small as pointed out by @gchanan .

I'd like quantize some of the important points raised by @soumith @t-vi into this thread for future reference.

# Summary of the problem

The main problem over here is that the `_version` of the `running_mean` and `running_var` is not changed once they go through a forward pass in the `train` mode. Due to this version counter not being bumped, the `gradcheck` is not able to compute the correct Jacobian matrix for the custom forward functions that we are specifying.

The `native_batch_norm` function is the function that implements the batch norm and can be found in the `native_functions.yaml` file with various implementations for CPU (in `CPUType.cpp`) and GPU.

# Possible solutions and tradeoffs

After spending a fair amount of time in understanding this issue, I figured that there might be two possible approaches to solving this issue:
1. Figure out a place where the `running_mean` and `running_var` tensors need to have their versions updated and insert some C++ code to make it happen.
2. Make changes to the declaration of `native_batch_norm` in `native_functions.yaml` such that it allows in-place (currently specified with `a!`) and optional (currently specific with `?`) arguments.

The first approach involves device-specific changes for every function that implements `native_batch_norm`, and is therefore not very clean since the `native_batch_norm` function is fully auto-generated and the only way to specify something about the arguments in a device-agnostic way is to do that within the `native_functions.yaml` file itself.

The second approach looks nice and simple, however there is currently no way to specify an argument as both optional and in-place. This requires somehow figuring out a way to specify both in-place and optional arguments, which requires a fair amount of tweaking with code generation and understanding how it might impact the JIT.

# Problems with second approach.

Currently making modifications to the argument inference is not possible since the autogen throws conflicts about type mismatch.

When making changes for this issue, it needs to be seen to that the batch norm optimizations in the JIT deal with the new calling signature (of in-place and optional together). Another problem is that since the arguments are not modified in-place in the inference mode, it is unclear whether marking the signature as 'in place modified' would wreak havoc with the JIT inference mode.

**Comment by v0dro** [comment]
@jjsjann123 so you suggest modifying the respective device-specific functions to bump version counter?

**Comment by csarofeen** [comment]
My understanding is we need to get rid of this in-place operation for JIT analysis. As we would otherwise need to build alias analysis which would be challenging. CC @zdevito @mruberry

**Comment by jjsjann123** [comment]
No, I was suggesting to remove the in-place update on `running_mean` and `running_var`, as @csarofeen pointed out.

Just a note on the proposed solution 1.
Updating version counter does not resolve potential bug on JIT. If we do not properly mark function schema for batch_norm (not putting in-place annotation `(!)` to the stats input), aliasing analysis would not identify them as input/output tensor rendering it possible for optimization passes to make un-safe optimization that change the semantics of the graph.

**Comment by albanD** [comment]
@jjsjann123 note that removing the inplace update means that we cannot use cudnn batchnorm anymore.

**Comment by jjsjann123** [comment]
Although it's not optimal, but we still can and Tensorflow does it that way.

We can have cudnn BN to run forward pass with 0 (or 1) momentum. Having a separate element-wise kernel to accumulate running_mean/running_var and then update it to the module.

**Comment by jjsjann123** [comment]
Or if we decide to keep the in-place update here.
I'm shamelessly plugin an ad for my issue proposing handling of aliasing information [still in draft and could use some reviews & comments]: https://github.com/pytorch/pytorch/issues/28191
😜

**Comment by albanD** [comment]
Given how complex modifying the codegen would be, a third option might be to write the batchnorm function binding by hand (based on the current codegened code) and add the proper version bumps.
We would need to make sure that the jit still handles it properly.
@gchanan would that be a good temporary solution until the codegen can be changed to support optional/inplace?

**Comment by gchanan** [comment]
@albanD: I mean, it's better than the status-quo, so I don't see why we shouldn't do it.

i.e. there are two problems:
1) we don't bump the version counters even though we actually update the tensors
2) the schema doesn't specify that the element is mutable.

maybe I'm missing something, but 2) seems like an issue whether or not you fix 1).  So I don't necessarily understand why we need to solve both problems at once.

**Comment by albanD** [comment]
Sounds good. 
So this split in 2 tasks where the first one is fairly small:

- Change the entry points of batchnorm in Aten/native to bump the version counter of the buffers when they are actually used.
- Change the codegen to support optional + inplace modified.

**Comment by gchanan** [comment]
`fixathon` for doing the first point only:
> Change the entry points of batchnorm in Aten/native to bump the version counter of the buffers when they are actually used.

**Comment by glaringlee** [comment]
Booked this as an onboarding task

**Comment by albanD** [comment]
There is a new update on this: Let's split the native function into two overloads to bypass this problem altogether.
@janeyx99 will take a look into this in the following weeks.

**Comment by jbschlosser** [comment]
Update: @janeyx99 looked into this and split `native_batch_norm` into two overloads.

@janeyx99 / @albanD am I correct that this would also have to be done for the backend-specific batch norm ops (e.g. `cudnn_batch_norm` / `miopen_batch_norm`)? In this case, I think we'd want to halt going down this route until the batch norm ops have been consolidated.

**Comment by janeyx99** [comment]
Update: I've split the native function into two private ops (with the correct schema) where this bug will no longer appear when one enables the python dispatcher (https://github.com/pytorch/pytorch/pull/88697). 

However, to fix this completely, we would have to do the following:
- Change the original native_batch_norm to be split into two schemas so the python dispatcher does not need to be enabled. This was not immediately done due to forwards and backwards incompatibility purposes with JIT tracing, and due to these same reasons, splitting the original native_batch_norm must wait a few weeks after #88697.
- There are FOUR other variations of batch_norm (like cudnn_batch_norm) that would need to receive the same treatment (and waiting time).

Instead of doing the same process for 4 other batch norms, we should just consolidate into one correct batch norm and dispatch from there.

**Comment by janeyx99** [comment]
Ah oops--was typing up my comment before seeing @jbschlosser's--but yes, absolutely, it is better to consolidate the ops instead of repeating the process.

**Comment by janeyx99** [comment]
@andrewor14 Does your work consolidating the batch_norms fix this issue completely?
